### PR TITLE
fix: Account for the expected kind when resolving turbofish generics

### LINKED ISSUE #5442: Turbofish for numeric generics in nested function call fails
### Aim

```rust
fn foo<let N: u32>() -> [u8; N] {
    [0; N]
}

fn bar<let N: u32>() -> [u8; N] {
    foo::<N>()
}

global M: u32 = 3;

fn main() {
    println(bar::<M>());
}
```

### Expected Behavior

Program to compile and execute successfully.

Note in the workaround that using turbofish for `bar` in `main` works.

### Bug

```bash
error: expected type, found numeric generic parameter N
  ┌─ /Users/michaelklein/Coding/noir/bug_from_regex/src/main.nr:6:11
  │
6 │     foo::<N>()
  │           - not a type
  │
```

### To Reproduce

1.
2.
3.
4.

### Project Impact

Blocker

### Impact Context

Medium-blocker: there's a lot of code that uses numeric generics and the workarounds are fairly heavy

### Workaround

Yes

### Workaround Description

Phantom data that's expected to compile to `()` can be passed around:

```rust
struct Phantom<let N: u32> { }

fn foo<let N: u32>(x: Phantom<N>) -> [u8; N] {
    [0; N]
}

fn bar<let N: u32>(x: Phantom<N>) -> [u8; N] {
    foo(x)
}

global M: u32 = 3;

fn main() {
    let phantom: Phantom<M> = Phantom {};
    println(bar::<M>(phantom));
}
```

### Additional Context

_No response_

### Installation Method

Compiled from source

### Nargo Version

nargo version = 0.31.0 noirc version = 0.31.0+b58a034a36dab6d17e10f2fcceab93e9073f53ed (git version hash: b58a034a36dab6d17e10f2fcceab93e9073f53ed, is dirty: false)

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #5442 

## Summary\*

In #5155 we make sure that we make sure we resolve against the correct type kind in various places such as when resolving trait bound generics or trait impl generics. This was not done for function calls. Now when resolving turbofish generics, rather than calling `resolve_type` in all cases aside type expressions, we check whether the variable we are elaborating is a function and fetch its already resolved generics and pass that in as the expected kind.

To keep the resolution logic the simple (I just zipped together the `direct_generics` from the function meta and the user specified turbofish generics), as `zip` uses the shortest iterator, I also added two checks for whether we have the correct turbofish count. We now have multiple of these checks rather than in just `instantiate`, but I felt it was the simplest logic and would lead to the least potential future confusion. 

I also have expanded the bug identified in the issue to also include turbofish generics for method calls. I have also included this as part of my test for this PR.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/tests.rs [hunk 1]
```diff
@@ -1455,6 +1455,79 @@ fn specify_method_types_with_turbofish() {
...
+            fn default() -> Self;
+        }
+
+        impl Default for Field {
+            fn default() -> Self { 0 }
+        }
+
+        // Need the above as we don't have access to the stdlib here.
+        // We also need to construct a concrete value of `U` without giving away its type
+        // as otherwise the unspecified type is ignored.
```
Comment (jfecher) [inline compiler/noirc_frontend/src/tests.rs:1506]: Another option in the future would be to write the builtin for the zeroed function. No need to change it for this PR though.

File: compiler/noirc_frontend/src/tests.rs [hunk 2]
```diff
@@ -1983,3 +2056,57 @@ fn underflowing_i8() {
...
+    }
+
+    global M: u32 = 3;
+
+    fn main() {
+        let _ = bar::<M>();
+    }
+    "#;
+    let errors = get_program_errors(src);
+    dbg!(errors.clone());
```
Comment (jfecher) [inline compiler/noirc_frontend/src/tests.rs:2110]:
```suggestion
    dbg!(&errors);
```
No need to clone here

File: compiler/noirc_frontend/src/elaborator/expressions.rs [hunk 3]
```diff
@@ -342,14 +326,36 @@ impl<'context> Elaborator<'context> {
                     }
                 };
 
-                if func_id != FuncId::dummy_id() {
+                // Perform any check that required information from the interned function.
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/expressions.rs:329]:
```suggestion
```
I don't think this comment is useful, "perform any check" is too broad for me and we already have a comment on resolving generics below.

File: compiler/noirc_frontend/src/elaborator/patterns.rs [hunk 4]
```diff
@@ -401,12 +401,45 @@ impl<'context> Elaborator<'context> {
...
+                        vecmap(generics_with_types, |(generic, unresolved_type)| {
+                            self.resolve_type_inner(unresolved_type, &generic.kind)
+                        })
+                    })
+                }
+                _ => None,
+            }
+        } else {
+            None
+        };
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/patterns.rs:441]:
```suggestion
        let generics = definition.and_then(|definition| {
            match &definition.kind {
                DefinitionKind::Function(function) => {
                    // Resolve generics using the expected kinds of the function we are calling
                    let direct_generics =
                        self.interner.function_meta(function).direct_generics.clone();

                    unresolved_turbofish.map(|option_inner| {
                        if option_inner.len() != direct_generics.len() {
                            let type_check_err = TypeCheckError::IncorrectTurbofishGenericCount {
                                expected_count: direct_generics.len(),
                                actual_count: option_inner.len(),
                                span,
                            };
                            self.push_err(type_check_err);
                        }

                        let generics_with_types = direct_generics.iter().zip(option_inner);
                        vecmap(generics_with_types, |(generic, unresolved_type)| {
                            self.resolve_type_inner(unresolved_type, &generic.kind)
                        })
                    })
                }
                _ => None,
            }
        });
```

File: compiler/noirc_frontend/src/elaborator/patterns.rs [hunk 5]
```diff
@@ -401,12 +401,45 @@ impl<'context> Elaborator<'context> {
...
+                                span,
+                            };
+                            self.push_err(type_check_err);
+                        }
+
+                        let generics_with_types = direct_generics.iter().zip(option_inner);
+                        vecmap(generics_with_types, |(generic, unresolved_type)| {
+                            self.resolve_type_inner(unresolved_type, &generic.kind)
+                        })
+                    })
```
Comment (jfecher) [inline compiler/noirc_frontend/src/elaborator/patterns.rs:435]: This code chunk seems to be repeated for method calls as well - we should have a helper for it

File: compiler/noirc_frontend/src/elaborator/patterns.rs (hunk 4 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/patterns.rs:441]: (only change is using `and_then`)

**Review by jfecher — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/tests.rs (hunk 2 above)
Comment (vezenovm) [reply compiler/noirc_frontend/src/tests.rs]: I ended up deleting this as the dbg is leftover from debugging.

File: compiler/noirc_frontend/src/elaborator/patterns.rs (hunk 5 above)
Comment (vezenovm) [reply compiler/noirc_frontend/src/elaborator/patterns.rs]: I have made a helper `resolve_turbofish_generics`

**Review by jfecher — APPROVED** [review]
:+1:

**Pull request merged**